kip-0026.md

```diff
@@ -0,0 +1,105 @@
+
+## Mnemonic Encoding
+
+We recommend following the adopted standard for encoding private keys to mnemonic seed phrases: [BIP-0039](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki) with 24 words, using the English BIP39 wordlist.
+
+We recommend not using a pass phrase to encrypt the mnemonic.
+
+# Rationale
+
+## Key Derivation
+
+Rationale for choosing SLIP-10:
+
+- SLIP-10 is the most widely adopted derivation standard and algorithm for ED25519 curve blockchain keypairs.
+- A custom or exotic derivation algorithm would have been a hindrance a contrained environments such as hardware wallets, where a non-standard algorithm will commonly require a firmware upgrade.
+- A standard derivation algorithm will reduce friction for multi-chain wallet integrations, as those will often support SLIP-10 derivation already, without the need to include (and audit) a custom library.
+
+The three-level derivation path `m/44'/626'/a'` was chosen after a number of considerations:
+
+- As simple as needs be, and no more: the fourth (change) and fifth (address index) levels of BIP-44 are superfluous for account-based blockchain keypairs.
+  - Including them could result in confusion about the correct level to iterate for generating additional accounts beyond the first, leading to a situation where only the first account is interoperable between wallets.
+  - Hardened vs not hardened: while SLIP-10 forbids non-hardened keys, there are libraries that do not enforce this. Including the last 2 BIP-44 levels would also create space for confusion between hardened and non-hardened path suffixes (e.g. `/0/0` and `/0'/0'`). This would produce different keys in wallets that accept the non-hardened derivation path, and the non-hardened paths would not be reproducible at all in wallets that use libraries that implement SLIP-10 strictly.
+- Widely used for account-based blockchains' ED25519 keypairs. See Stellar, NEM, Monero and Tezos in the [list of Trezor BIP-44 derivation paths](https://docs.trezor.io/trezor-firmware/misc/coins-bip44-paths.html#list-of-used-derivation-paths).
+- SLIP-10 compliant. BIP-44 specifies 5 levels, but SLIP-10 does not recommend a specific number of levels.
+- `626` is the [SLIP-44](https://github.com/satoshilabs/slips/blob/master/slip-0044.md) coin type identifier for Kadena.
+
+## Mnemonic Encoding
+
+BIP-39 is nearly universally adopted as the mnemonic encoding method.
+
+24 word mnemonics: support 256 bits of entropy.
+
+The English BIP39 wordlist is recommended for interoperability.
+
+Empty passphrase:
+
+- Allows interoperability.
+  - Few wallets implement encrypted mnemonic recovery.
+- Encrypting a mnemonic with a passphrase may provide a sense of security, but:
+  - A leaked encrypted mnemonic phrase can be quickly brute-forced offline ([BIP-39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#from-mnemonic-to-seed): 2048 round PBKDF2 using salted HMAC-SHA512).
```

Comment:
:exclamation: :question:  :open_mouth:
Are you sure that a (good) password derived with a SHA256-2048 PBKDF2 can be **quickly** brute-forced ?

Reply:
more specifically the assertion that it can be quickly brute-forced is not backed by the reference as far as I can see.

Reply:
Perhaps "significant speed" is more accurate than "quickly", but I think the essence of the statement holds.

A _good_ password could withstand MD5, but it would probably end up looking like another seed phrase to be stored securely, away from the encrypted seedphrase. An average user password would succumb reasonably quickly, especially if the seed was compromised on their machine, where a lot of information could be used to build up a dictionary.

For the current BIP39 spec of HMAC-SHA-512 [1] I estimate that an off-the-shelf 4090 can do about ~1521 kH/s[2] which would map to:

- rockyou.txt wordlist (14M): 9 seconds
- crackstation wordlist (1.2B): 13 minutes
- crackstation * 100 masks (121B): 22 hours
- bruteforce alpha/lower 8 letters (208B): 38 hours
- bruteforce alpha/lower + numeric 8 letters (2.8T): 21 days

And so on. It is not exactly a done-deal given enough entropy, but I'd still call it a fast enough attack with consumer hardware, which will only become cheaper and more accessible in the future[3][4].

I would consider "slow" here to be something like LUKS2/argon2i which is molasses-slow to dictionary attack or bruteforce, or pbkdf2/sha512 with a very large number of rounds (600K+.)

All that said, it should probably read "bruteforced with significant speed", instead of "quickly", so I will change that.

[1] HMAC-SHA-256 would be even faster: A 4090 should do 2048 rounds @ [~4170 kH/s](https://gist.github.com/Chick3nman/32e662a5bb63bc4f51b847bb422222fd#file-rtx_4090_v6-2-6-benchmark-L1592)

[2] Extrapolating from the speed given for 999 iterations [here](https://gist.github.com/Chick3nman/32e662a5bb63bc4f51b847bb422222fd#file-rtx_4090_v6-2-6-benchmark-L1250) of 3120.9 kH/s - rounds to speed should be ~linear

[3] Consider the lastpass situation where some of their early accounts were stuck at < 10K iterations and were cracked after a disasterous leak. Not sure what their hashing function is though.

[4] We don't even know what the "cracking station" state of the art is at the moment. It would not surprise me to learn that there are GPU farms or even custom ASICs that can smoke these numbers.